virtcontainers/device/drivers/block.go

```diff
@@ -88,53 +78,68 @@ func (device *BlockDevice) Attach(devReceiver api.DeviceReceiver) (err error) {
 		return err
 	}
 
-	drive := Drive{
-		File:   device.DeviceInfo.HostPath,
-		Format: "raw",
-		ID:     utils.MakeNameID("drive", device.DeviceInfo.ID, maxDevIDSize),
-		Index:  index,
-	}
-
-	deviceLogger().WithField("device", device.DeviceInfo.HostPath).Info("Attaching block device")
-	device.BlockDrive = &drive
-	if err = devReceiver.HotplugAddDevice(device, config.DeviceBlock); err != nil {
-		return err
+	drive := &config.BlockDrive{
+		File:          device.DeviceInfo.HostPath,
+		Format:        "raw",
+		ID:            utils.MakeNameID("drive", device.DeviceInfo.ID, maxDevIDSize),
+		Index:         index,
+		ContainerPath: device.DeviceInfo.ContainerPath,
 	}
 
-	device.DeviceInfo.Hotplugged = true
-
 	driveName, err := utils.GetVirtDriveName(index)
 	if err != nil {
 		return err
 	}
 
 	customOptions := device.DeviceInfo.DriverOptions
 	if customOptions != nil && customOptions["block-driver"] == "virtio-blk" {
-		device.VirtPath = filepath.Join("/dev", driveName)
-		device.PCIAddr = drive.PCIAddr
+		drive.VirtPath = filepath.Join("/dev", driveName)
 	} else {
 		scsiAddr, err := utils.GetSCSIAddress(index)
 		if err != nil {
 			return err
 		}
 
-		device.SCSIAddr = scsiAddr
+		drive.SCSIAddr = scsiAddr
+	}
+
+	deviceLogger().WithField("device", device.DeviceInfo.HostPath).Info("Attaching block device")
+	device.BlockDrive = drive
+	if err = devReceiver.HotplugAddDevice(device, config.DeviceBlock); err != nil {
+		return err
 	}
 
+	device.DeviceInfo.Hotplugged = true
+
 	return nil
 }
 
 // Detach is standard interface of api.Device, it's used to remove device from some
 // DeviceReceiver
 func (device *BlockDevice) Detach(devReceiver api.DeviceReceiver) error {
-	if device.DeviceInfo.Hotplugged {
-		deviceLogger().WithField("device", device.DeviceInfo.HostPath).Info("Unplugging block device")
-
-		if err := devReceiver.HotplugRemoveDevice(device, config.DeviceBlock); err != nil {
-			deviceLogger().WithError(err).Error("Failed to unplug block device")
-			return err
+	device.DevManager.Lock()
+	defer device.DevManager.Unlock()
+
+	// make sure this is the last device we are referencing
+	attached := device.DevManager.AttachedDevices(device)
+	switch len(attached) {
+	case 0:
+		// this should never happen, report error if it occurs
```

Comment:
Could you add a `return fmt.Errorf("BUG: ...")` or similar?

Reply:
Addressed: and find a wrong test case via fixing this!